ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/ConflictCheckingLedgerBridge.scala

```diff
@@ -0,0 +1,623 @@
+      )
+    )
+
+    queueSource.merge(source)
+  }
+
+  val (queue: BoundedSourceQueue[Submission], queueSource: Source[(Offset, Update), NotUsed]) =
+    Source
+      .queue[Submission](submissionBufferSize)
+      .map { submission =>
+        val newOffsetIdx = offsetIdx.incrementAndGet()
+        val newOffset = toOffset(newOffsetIdx)
+        (newOffset, successMapper(submission, newOffsetIdx, participantId))
+      }
+      .preMaterialize()
+
+  logger.info(
+    s"BridgeLedgerFactory initialized. Configuration: [maxDedupSeconds: $maxDedupSeconds, submissionBufferSize: $submissionBufferSize]"
+  )
+
+  private def submit(submission: Submission): CompletionStage[SubmissionResult] =
+    toSubmissionResult(queue.offer(submission))
+
+  private def precomputeTransactionOutputs(
+      transaction: Transaction
+  ): Future[Either[SoxRejection, TransactionEffects]] = {
+    lazy val contractKeyInputs =
+      transaction.transaction.contractKeyInputs.left.map(invalidInputFromParticipant(transaction))
+    lazy val inputContracts = transaction.transaction.inputContracts
+    lazy val updatedContractKeys = transaction.transaction.updatedContractKeys
+    lazy val consumedContracts = transaction.transaction.consumedContracts
+
+    Timed.future(
+      metrics.daml.SoX.precomputeTransactionOutputs,
+      Future(contractKeyInputs).map(
+        _.map((_, inputContracts, updatedContractKeys, consumedContracts))
+      ),
+    )
+  }
+
+  private def conflictCheckWithDelta
+      : () => Either[SoxRejection, (Offset, Transaction, TransactionEffects)] => Iterable[
+        (Offset, Update)
+      ] = () => {
+    @volatile var sequencerQueueState: SequencerState = SequencerState()(metrics)
+
+    {
+      case Left(rejection) =>
+        Timed.value(
+          metrics.daml.SoX.sequenceDuration, {
+            val newIndex = offsetIdx.incrementAndGet()
+            val newOffset = toOffset(newIndex)
+            Iterable(newOffset -> toRejection(rejection))
+          },
+        )
+      case Right(
+            (
+              noConflictUpTo,
+              transaction,
+              (keyInputs, inputContracts, updatedKeys, consumedContracts),
+            )
+          ) =>
+        Timed.value(
+          metrics.daml.SoX.sequenceDuration, {
+            val newIndex = offsetIdx.incrementAndGet()
+            val newOffset = toOffset(newIndex)
+            conflictCheckWithInTransit(
+              sequencerQueueState,
+              keyInputs,
+              inputContracts,
+              transaction,
+            ) match {
+              case Left(rejection) => Iterable(newOffset -> toRejection(rejection))
+              case Right(acceptedTx) =>
+                sequencerQueueState = sequencerQueueState
+                  .dequeue(noConflictUpTo)
```

Comment:
Is it important to make the conflict checking against the minimal state?
If not how about removing the dequeue (eviction) from the sequential pipe-line, and doing eviction separately?
This would have the beneif that we do not have to tdo the expensive eviction all the time, just regularly.
This could be achieved via tracking the noConflictUpTo in some external state, as well as the SequencerState (noConflictUpTo could be part of SequencerState as well to be self-contained), and a separate process is doing cleanup/eviciont